proposals/0000-compiler-settings.md

```diff
@@ -0,0 +1,279 @@
+is left in an intermediate, partially updated state or the module being updated
+on a branch that is merged once all updates have been completed. In the former
+case, it is easy to introduce new warnings as new code is added to the codebase
+since only warnings are being used. In the later case, the branch must be kept
+up to date in the face of changes being made to the codebase on main, forcing
+one to continually need to update the branch to resolve merge conflicts against
+main.
+
+In contrast by allowing for command line flags to be applied one file at a time,
+users can update code using a warningsAsErrors based migration path. This
+involves instead:
+
+1. Enabling warningsAsErrors for the new feature on the specific file.
+
+2. Fix all of the errors in the file.
+
+3. Commit the updated file into main.
+
+Since one is only updating one file at a time and using warningsAsErrors, the
+updates can be done very incrementally, backsliding cannot happen, and main is
+never in an intermediate, partially updated state.
+
+## Proposed solution
+
+We propose introducing a new top level statement called `compilerSettings` that
+takes a list of enums of type `CompilerSetting`. Each enum case'a associated
+values represent arguments that would be passed to the compiler flag on the
+command line. The compiler upon parsing a `compilerSettings` statement updates
+its internal state to set the appropriate settings before compiling the rest of
+the file. An example of such a `compilerSettings` statement is the following:
+
+```swift
+compilerSettings [
+    .warningAsError(.concurrency),
+    .strictConcurrency(.complete),
+    .enableUpcomingFeature(.existentialAny),
+    .defaultIsolation(MainActor.self),
+]
+```
+
+By specifying compiler options in such a manner, we are able to solve all of the
+problems above:
+
+1. The programmer can avoid module splitting when they wish to use compiler
+   flags only on a specific file.
+
+2. A programmer can specify options on a per file basis instead of only on a per
+   module basis allowing for warningsAsError to be used per file instead of per
+   module migration to use new features.
+
+As an additional benefit since enums are being used, code completion can make
+the compiler options discoverable to the user in the IDE.
+
+## Detailed design
+
+We will update the swift grammer to allow for a new top level statement called
+`compilerSettings` that takes a list of expressions:
+
+```text
+top-level-statement ::= 'compilerSettings' '[' (expr ',')* expr ','? ']'
+```
+
+During typechecking, we will validate that each of the expressions are instances
+of the enum `CompilerSetting`. `CompilerSetting` will be a resilient enum
+defined in the standard library that provides the following API:
+
+```
+public enum CompilerSetting {
+case enableUpcomingFeature(UpcomingFeatureCompilerSetting)
+case enableExperimentalFeature(ExperimentalFeatureCompilerSetting)
+case warningAsError(WarningAsErrorCompilerSetting)
```

Comment:
This is interesting. So far everyone defines this on their CI. If you start defining this on a per file basis it might be that a compiler introduces a new warning and suddenly your dependency fails to build since they added this to their per-file compiler setting.